src/query.rs

```diff
@@ -701,6 +737,15 @@ macro_rules! tuple_impl {
                 $($name::release(archetype);)*
             }
 
+            #[allow(unused_variables, unused_mut)]
+            fn for_each_borrow(mut f: impl FnMut(TypeId, usize, bool)) {
+                let mut counter = 0;
```

Comment:
Coming the borrow checking versus checking uniqueness below, wouldn't that counter need to be used for the whole type hierarchy? I think this way, we'll get examples like
```
((&i32, &mut i32), (&i32, &mut i32)) => ((0,1), (0,1))
```
instead of
```
((&u64, &mut i32), (&u64, &mut i32)) => ((0,1), (2,3))
```
and the above my actually check because `&mut i32` is positioned at `1` in both instances?

Reply:
The callback is invoked for each individual borrow, and not for intermediate constructions like tuples. When tuples are nested, you end up with multiple counters, but an outer counter always ignores the output from the inner; you can see below that we discard the counter value passed in by the inner type. As a result, we should always get the second case.

Reply:
Maybe it would be simplest to not do any counting within `for_each_borrow` at all just keep the `i` and `j` counters as mutable variables on the stack of `QueryMut::new`? If `Q::Fetch::for_each_borrow` is guaranteed to always produce the same visitation order, keeping count could be done outside of it entirely.

Reply:
> but an outer counter always ignores the output from the inner; you can see below that we discard the counter value passed in by the inner type.

Yes, now that you I mention it, I do see it. I would still argue that keeping the counters on the outer most layer in `QueryMut::new` would actually be the simplest way as it would avoid having to ignore any layer of counting.

Reply:
Good catch; fixed.